File: week 4/inversions/inversions.py
def merge(a, b, left, ave, right):
    i, j, k = left, ave, left
    count = 0
    while (i <= ave -1 and j <= right):
        if (a[i] <= a[j]):
            b[k] = a[i]
            i += 1
            k += 1
        else:
            b[k] = a[j]
            count += ave - i
            j += 1
            k += 1
    while (i <= ave - 1):
        b[k] = a[i]
        i += 1
        k += 1
    while (j <= right):
        b[k] = a[j]
        j += 1
        k += 1
    for i in xrange(left,right+1):
        a[i] = b[i]
    return count
			
def get_number_of_inversions(a, b, left, right):
    number_of_inversions = 0
    if right <= left:
        return number_of_inversions
    ave = (left + right) // 2
    number_of_inversions += get_number_of_inversions(a, b, left, ave)
    number_of_inversions += get_number_of_inversions(a, b, ave+1, right)
    number_of_inversions += merge(a, b, left, ave+1, right)
    return number_of_inversions

if __name__ == '__main__':
    n = int(input())
    a = list(map(int, raw_input().split()))
    b = n * [0]
    print(get_number_of_inversions(a, b, 0, len(a)-1))

File: week 4/inversions/inversions.cpp
#include <iostream>
#include <vector>

using std::vector;

long long merge(vector<int> &a, vector<int> &b, size_t left, size_t ave, size_t right) {
	    size_t i = left, j = ave, k = left;
        long long count = 0;
        while (i <= ave -1 && j <= right) {
            if (a[i] <= a[j]) {
                b[k] = a[i];
				i++;
				k++;
            } else {
                b[k] = a[j];
				count += ave - i;
				j++;
				k++;
            }
        }
        while (i <= ave - 1) {
            b[k] = a[i];
            i++;
            k++;
        }
        while (j <= right) {
            b[k] = a[j];
            j++;
            k++;
        }	
		for (i=left; i <= right; i++)
			a[i] = b[i];
        return count;
}

long long get_number_of_inversions(vector<int> &a, vector<int> &b, size_t left, size_t right) {
  long long number_of_inversions = 0;
  if (right <= left) return number_of_inversions;
  size_t ave = left + (right - left) / 2;
  number_of_inversions += get_number_of_inversions(a, b, left, ave);
  number_of_inversions += get_number_of_inversions(a, b, ave+1, right);
  number_of_inversions += merge(a, b, left, ave+1, right); 
  return number_of_inversions;
}

int main() {
  int n;
  std::cin >> n;
  vector<int> a(n);
  for (size_t i = 0; i < a.size(); i++) {
    std::cin >> a[i];
  }
  vector<int> b(a.size());
  std::cout << get_number_of_inversions(a, b, 0, a.size()-1) << '\n';
}
